libphal/phal_sbedump.C

```diff
@@ -0,0 +1,228 @@
+#include <ekb/chips/p10/procedures/hwp/lib/pibms_regs2dump.H>
+#include <ekb/hwpf/fapi2/include/return_code_defs.H>
+#include <fapi2.H>
+
+#include <filesystem>
+#include <fstream>
+#include <string>
+#include <vector>
+
+namespace openpower::phal
+{
+namespace dump
+{
+
+using namespace openpower::phal::logging;
+
+void preCollection(struct pdbg_target* proc)
+{
+	struct pdbg_target *fsi, *pib;
+	char path[16];
+	sprintf(path, "/proc%d/fsi", pdbg_target_index(proc));
+	fsi = pdbg_target_from_path(NULL, path);
+	assert(fsi);
+	sprintf(path, "/proc%d/pib", pdbg_target_index(proc));
+	pib = pdbg_target_from_path(NULL, path);
+	assert(pib);
+
+	if ((pdbg_target_probe(fsi) != PDBG_TARGET_ENABLED ||
+	     pdbg_target_probe(pib) != PDBG_TARGET_ENABLED)) {
+		pdbg_target_status_set(proc, PDBG_TARGET_DISABLED);
+	}
+}
+
+int writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
+{
+	int rc = -1;
+	std::ofstream outfile{dumpPath, std::ios::out | std::ios::binary};
+	if (!outfile.good()) {
+		int err = errno;
+		log(level::INFO,
+		    "Failed to open the dump file for writing err=%d", err);
+		return rc;
+	}
+	outfile.exceptions(std::ifstream::failbit | std::ifstream::badbit |
+			   std::ifstream::eofbit);
+	try {
+		outfile.write(data, len);
+	} catch (std::ofstream::failure& oe) {
+		int err = errno;
+		log(level::INFO,
+		    "Failed to write to dump file err=%d error message=%s", err,
+		    oe.what());
+		return rc;
+	}
+	outfile.close();
+	return 0;
+}
+
+int collectSBEDump(uint32_t id, uint8_t failingUnit,
+		   std::filesystem::path& dumpPath)
+{
+
+	int rc = -1;
+	log(level::INFO,
+	    "Collecting SBE dump: path=%s, id=%d, chip position=%d",
+	    dumpPath.string().c_str(), id, failingUnit);
+	pdbg_set_backend(PDBG_BACKEND_KERNEL, NULL);
+
+	if (!pdbg_targets_init(NULL)) {
+		log(level::ERROR, "pdbg_targets_init failed");
+		throw std::runtime_error("pdbg target initialization failed");
+	}
+	pdbg_set_loglevel(PDBG_INFO);
+
+	struct pdbg_target* target;
+	struct pdbg_target* proc = NULL;
+	pdbg_for_each_class_target("proc", target)
+	{
+		if (pdbg_target_probe(target) != PDBG_TARGET_ENABLED) {
+			continue;
+		}
+
+		if (pdbg_target_index(proc) == failingUnit) {
+			proc = target;
+		}
+	}
+	if (proc == NULL) {
+		log(level::ERROR, "No Proc target found to execute the dump");
+		return rc;
+	}
+
+	std::stringstream ss;
+	ss << std::setw(8) << std::setfill('0') << id;
+	std::string idStr = ss.str();
+
+	std::string baseFilename =
+	    idStr + ".0_" + std::to_string(failingUnit) + "_SbeData_p10_";
+	preCollection(proc);
+	fapi2::ReturnCode fapirc;
+
+	// Collect SBE local register dump
+	std::vector<SBESCOMRegValue_t> sbeScomRegValue;
+	fapirc = p10_sbe_localreg_dump(proc, true, sbeScomRegValue);
+	if (fapirc != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed in p10_sbe_localreg_dump for proc=%s, rc=0x%08X",
+		    pdbg_target_path(proc), fapirc);
+	}
```

Comment:
Looks like dump hwp desin't support error xmls. Since the dump collection is triggerd during error path , i think journal log is suffient for any hwp , not sure any value add of creating error log. If error log is required , we may need callback mechansim to log error .